### ISSUE TITLE
Can't differentiate through indexing operation

### ISSUE THREAD

**Issue by DavidMinorNV** [pytorch/pytorch#121971]
### 🐛 Describe the bug

I might as well flag this, although I think the fix is just a case of implementing a symbolic override for aten::_index_put_impl correctly. Here's a test case:

```
import torch


class Model(torch.nn.Module):

    def forward(self, inputs):
        pred = inputs[0]
        pred = pred.detach().requires_grad_()
        with torch.enable_grad():
            
            loss_tot = pred[..., [0]].sum()
            grad_outs = [loss_tot]
            grad_ins = [pred]
            return torch.autograd.grad(grad_outs, grad_ins)[0]


mdl = Model()

print("export model")
torch.onnx.export(
    Model(),
    [torch.randn([20, 2, 256])],
    "modelthing.onnx",
    input_names=["xseq"],
    opset_version=17,
    output_names=["lossgrad"],
    verbose=True
)
```

If you run that, you get the following error:

```
torch.onnx.errors.UnsupportedOperatorError: Exporting the operator 'aten::_index_put_impl' to ONNX opset version 17 is not supported. Please feel free to request support or submit a pull request on PyTorch GitHub: https://github.com/pytorch/pytorch/issues.
```

### Versions

PyTorch version: 2.2.1+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Enterprise
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.25.1
Libc version: N/A

Python version: 3.10.11 (tags/v3.10.11:7d4cc5a, Apr 5 2023, 00:38:17) [MSC v.1929 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19045-SP0
Is CUDA available: True
CUDA runtime version: 10.0.130
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090 Ti
Nvidia driver version: 536.23
cuDNN version: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.8\bin\cudnn_ops_train64_8.dll
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture=9
CurrentClockSpeed=3501
DeviceID=CPU0
Family=107
L2CacheSize=16384
L2CacheSpeed=
Manufacturer=AuthenticAMD
MaxClockSpeed=3501
Name=AMD Ryzen Threadripper PRO 3975WX 32-Cores
ProcessorType=3
Revision=12544

Versions of relevant libraries:
[pip3] functorch==2.0.0
[pip3] lovely-numpy==0.2.8
[pip3] numpy==1.24.3
[pip3] onnx==1.14.1
[pip3] onnx-graphsurgeon==0.3.27
[pip3] onnxconverter-common==1.13.0
[pip3] onnxruntime==1.15.1
[pip3] optree==0.10.0
[pip3] pytorch-lightning==1.4.2
[pip3] tf2onnx==1.16.0
[pip3] torch==2.2.1+cu118
[pip3] torch-cluster==1.6.1
[pip3] torch-fidelity==0.3.0
[pip3] torch-geometric==2.3.0
[pip3] torch-scatter==2.1.1
[pip3] torch-sparse==0.6.17
[pip3] torch-spline-conv==1.2.2
[pip3] torchaudio==2.2.1+cu118
[pip3] torchmetrics==0.6.0
[pip3] torchvision==0.17.1+cu118
[conda] blas 1.0 mkl
[conda] cudatoolkit 11.6.0 hc0ea762_10 conda-forge
[conda] libblas 3.9.0 16_win64_mkl conda-forge
[conda] libcblas 3.9.0 16_win64_mkl conda-forge
[conda] liblapack 3.9.0 16_win64_mkl conda-forge
[conda] mkl 2022.1.0 h6a75c08_874 conda-forge
[conda] mkl-include 2023.2.0 intel_49496 intel
[conda] mkl-static 2023.2.0 intel_49496 intel
[conda] numpy 1.24.2 py310hd02465a_0 conda-forge
[conda] pytorch 1.12.0 py3.10_cuda11.6_cudnn8_0 pytorch
[conda] pytorch-mutex 1.0 cuda pytorch

**Comment by thiagocrepaldi** [comment]
Hi David, we don't support this symbolic and we don't have plans to support this.

Please try the new ONNX exporter and reopen this issue if it also doesn't work for you: [quick torch.onnx.dynamo_export API tutorial](https://pytorch.org/tutorials/beginner/onnx/export_simple_model_to_onnx_tutorial.html)

Another idea is to contribute a PR for _inxdex_put_impl to the project yourself :)

**Comment by DavidMinorNV** [comment]
I don't know if this appears outside the context of backpropagation actually, which is completely unsupported by dynamo_export... I think my fix was just forwarding to the implementation of aten::index_put

By the way, I have a "pull request" with all the workarounds I had to do to get my backprop onnx to export here (including that one):

https://github.com/pytorch/pytorch/pull/124677

I say "pull request" in quotations because I have no intention of merging it and I'd like to discuss the next steps for getting this supported properly. Would you be able to put me in touch with anyone who can help me here? I don't want to have to maintain that branch forever and I'd like to try and make some progress doing this officially while it's still fresh in my memory.

**Comment by thiagocrepaldi** [comment]
feel free to convert that into an actual PR (not draft) and I people on reviewers field will look into it. i added the relevant labels

**Comment by DavidMinorNV** [comment]
Oh ok. I mean like I say, it's not really intended as an actual PR because it's really just a bunch of workarounds, and if people get all Linus Torvalds on me I will become The Joker. As long as that's understood though I can remove the draft flag if that makes things easier...

I imagine this will be a long discussion - I've tried to document all the things I did to just brute force it into working and most of them are bad. Maybe there are alternative/less dangerous ways of massaging things so they actually work via that path though. I guess ideally we'd make a start on supporting backprop in the dynamo_export path too.

Thanks
